### PR TITLE
Implement libinternals

### PR DESCRIPTION
**Pull request by jbrockmendel**
A bunch of functions from `lib` that are only used in core.internals, have light dependencies.

Small optimizations: made `slice_getitem` a `cdef` instead of `def`, use PySlice_Check instead of isinstance.

### CODE REVIEW

**Comment by jreback** [comment]
pls just do a move. do a followup to change things.

**Comment by jbrockmendel** [comment]
OK, just changed to be pure cut/paste, will do optimizations in follow-up.

File: pandas/_libs/internals.pyx [hunk 1]
```diff
@@ -0,0 +1,438 @@
...
+
+cdef extern from "compat_helper.h":
+    cdef int slice_get_indices(PyObject* s, Py_ssize_t length,
+                               Py_ssize_t *start, Py_ssize_t *stop,
+                               Py_ssize_t *step,
+                               Py_ssize_t *slicelength) except -1
+
+
+cdef class BlockPlacement:
+    # __slots__ = '_as_slice', '_as_array', '_len'
```
Comment (jreback) [inline pandas/_libs/internals.pyx]: in followup is fine, pls add doc-strings

File: pandas/_libs/internals.pyx [hunk 2]
```diff
@@ -0,0 +1,438 @@
...
+        return BlockPlacement(np.delete(self.as_array, loc, axis=0))
+
+    def append(self, others):
+        if len(others) == 0:
+            return self
+
+        return BlockPlacement(np.concatenate([self.as_array] +
+                                             [o.as_array for o in others]))
+
+    cdef iadd(self, other):
```
Comment (jreback) [inline pandas/_libs/internals.pyx]: these should prob be typed

**Review by jreback — APPROVED** [review]
this is fine. waiting on travis. certainly take optimzations (pls back it up with a timeit run as you asv seem broken :<). and doc-strings in a followup.

**Comment by jreback** [comment]
shouldn't this be called internals and not libinternals?

**Comment by jbrockmendel** [comment]
> shouldn't this be called internals and not libinternals?

The filename is "internals.pyx".  I've gotten used to the convention `import pandas._libs.internals as libinternals`

File: pandas/_libs/internals.pyx (hunk 2 above)
Comment (jbrockmendel) [reply pandas/_libs/internals.pyx]: Yep.  Bunch of small things like this I'll do in the next pass.  For now it's pure cut/paste.

**Comment by jreback** [comment]
thanks!

**Comment by jreback** [comment]
follow up for docs and small optimizations ok

**Comment by jbrockmendel** [comment]
Huh? Was this meant to be closed or merged?

**Comment by jreback** [comment]
it is merged 

see the hash above

**Pull request closed without merging**